### PR TITLE
Add predefined ``timestep()`` function

### PR DESCRIPTION
**Pull request by clinssen**
Add a ``timestep()`` function so models can be formulated in more generic terms than using the ``resolution()`` function. The latter assumes a fixed simulation timestep; the former is compatible with timesteps of arbitrary size (with subsequent steps potentially being of different size).

Existing time-based functions (``resolution()`` and ``steps()``) will remain, in order to be able to formulate models optimised for fixed-timestep simulators. This is also necessary to get a numerically identical result with NEST for at least one of the tests (stdp triplet synapse test). A warning will be emitted during model validation with a recommendation to use ``timestep()`` where possible.

New refractoriness mechanisms are implemented based on a countdown ODE (thanks to Tom Tetzlaff and Markus Diesmann!) This means that neither ``resolution()`` nor ``timestep()`` needs to be used any longer in the neuron models for the refractoriness mechanism. (See discussion in #988).

As a sidegoal: the triplet STDP synapse test contained simultaneously occurring pre- and post-spikes in the test set, which is not yet properly handled (this is already spun off to https://github.com/nest/nestml/issues/703). This meant that passing the test was relying on poorly predictable floating point behaviour. I have updated the spike sequences to not contain overlapping spikes, set the proper edge case flags and enabled the test for multiple synaptic delays (which now works).

## TODO
- [x] Change synapse models in the Jupyter notebooks

### CODE REVIEW

File: models/neurons/aeif_cond_alpha_neuron.nestml [hunk 1]
```diff
@@ -105,23 +106,16 @@ model aeif_cond_alpha_neuron:
...
-            refr_t -= resolution()
+            integrate_odes(w, refr_t)
         else:
-            # neuron not refractory, so evolve all ODEs (including V_m)
-            integrate_odes()
+            # neuron not refractory
+            integrate_odes(V_m, w)
 
-    onCondition(V_m >= V_peak):
+    onCondition(V_m >= V_peak and refr_t <= 0 ms):
```
Comment (tomtetzlaff) [inline models/neurons/aeif_cond_alpha_neuron.nestml:116]: Here and in all other neuron models with a similar construct, I'd swap the order of conditions: 
` onCondition(refr_t <= 0 ms and V_m >= V_peak)`

File: models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+default. If the property ``with_refr_input`` is set to true, such
+spikes are added to the membrane potential at the end of the
+refractory period, dampened according to the interval between
+arrival and end of refractoriness.
+
+The general framework for the consistent formulation of systems with
+neuron like dynamics interacting by point events is described in
+[1]_.  A flow chart can be found in [2]_.
+
+This model differs from ``iaf_psc_delta_fixed_timestep`` in that it
```
Comment (tomtetzlaff) [inline models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml:24]: This is certainly a typo and should read:
`This model differs from ``iaf_psc_delta`` ...`

File: models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    input:
+        spikes <- spike
+        I_stim pA <- continuous
+
+    output:
+        spike
+
+    update:
+        if is_refractory:
```
Comment (tomtetzlaff) [inline models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml:78]: This could be simplified by removing the boolean state variable `is_refractory`, and replacing

`onCondition(not is_refractory)` 

by 

`onCondition(refr_t <= 0 ms)`

(and similarly for all other refractoriness conditions)

File: models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        # threshold crossing
+        V_m = V_reset
+
+        if refr_T > 0 ms:
+            refr_t = refr_T    # start of the refractory period
+            is_refractory = true
+
+        emit_spike()
+
+    onCondition(is_refractory and refr_t <= resolution() / 2):
```
Comment (tomtetzlaff) [inline models/neurons/iaf_psc_delta_fixed_timestep_neuron.nestml:95]: Is this block really necessart?

File: doc/nestml_language/neurons_in_nestml.rst [hunk 5]
```diff
@@ -260,13 +260,13 @@ In order to model an absolute refractory state, in which the neuron cannot fire
    update:
        if is_refractory:
            # neuron is absolute refractory, do not evolve V_m
-           refr_t -= resolution()
```
Comment (tomtetzlaff) [inline doc/nestml_language/neurons_in_nestml.rst:263]: Shouldn't the entire section "Implementing refractoriness" be updated to explain the new way of handling refractoriness?

**Review by tomtetzlaff — CHANGES_REQUESTED** [review]
I have reviewed the part concerned with the ODE-based refractoriness implementation. Apart from few minor things (see detailed comments), this looks very good. Thanks for implementing this idea.

**Comment by clinssen** [comment]
Thank you for the detailed review! I have implemented all your comments, especially having done a rewrite of the "Implementing refractoriness" section in the neuron models documentation. Could you have another look? Much obliged!

**Review by tomtetzlaff — APPROVED** [review]
Looks very good and consistent now. Thanks a lot.

File: tests/nest_tests/stdp_triplet_synapse_test.py [hunk 6]
```diff
@@ -470,32 +456,19 @@ def _test_stdp_triplet_synapse(delay, spike_times_len):
...
 def test_stdp_triplet_synapse_delay_5(spike_times_len):
-    logging.warning("XXX: TODO: this test is disabled due to https://github.com/nest/nestml/issues/703")
-    # delay = 5.
-    # _test_stdp_triplet_synapse(delay, spike_times_len)
-
-
-logging.warning("XXX: TODO: xfail test due to https://github.com/nest/nestml/issues/703")
-# @pytest.mark.xfail(strict=True, raises=Exception)
-# @pytest.mark.parametrize("spike_times_len", [1, 10, 100])
+    delay = 5.
```
Comment (pnbabu) [inline tests/nest_tests/stdp_triplet_synapse_test.py:467]: Could you parametrize the `delay` as well?

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: tests/nest_tests/stdp_triplet_synapse_test.py (hunk 6 above)
Comment (clinssen) [reply tests/nest_tests/stdp_triplet_synapse_test.py:467]: Done!

File: doc/nestml_language/neurons_in_nestml.rst [hunk 7]
```diff
@@ -242,31 +242,54 @@ Output
...
+       emit_spike()
+
+The disadvantage of this method is that it requires a call to the ``resolution()`` function, which is only supported by fixed-timestep simulators. To write the model in a more generic way, the refractoriness timer can alternatively be expressed as an ODE:
+
+.. code-block:: nestml
 
    equations:
+       refr_t' = -1 / s    # a timer counting back down to zero
+
+Typically, the membrane potential should remain clamped to the reset or leak potential during the refractory period. It depends on the intended behavior of the model whether the synaptic currents and conductances also continue to be integrated or whether they are reset, and whether incoming spikes during the refractory period are taking into account or ignored.
```
[the remark this answers is not in the document]
Comment (clinssen) [reply doc/nestml_language/neurons_in_nestml.rst:269]: Fixed!
